File: demo/src/main/java/com/example/demo/service/CarService.java
package com.example.demo.service;

import com.example.demo.dto.CarDto;
import com.example.demo.mapper.CarMapper;
import com.example.demo.model.Car;
import com.example.demo.model.Client;
import com.example.demo.model.OrderEntity;
import com.example.demo.repository.CarRepository;
import com.example.demo.repository.ClientRepository;
import jakarta.transaction.Transactional;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;
import org.springframework.data.domain.Sort;

import java.time.LocalDate;
import java.util.Collections;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

@Service
public class CarService {

    @Autowired
    private CarRepository carRepository;

    @Autowired
    private ClientRepository clientRepository;

    public List<CarDto> getAllCars() {
        return carRepository.findAll(Sort.by(Sort.Direction.ASC, "id")).stream()
                .map(CarMapper::toDto)
                .collect(Collectors.toList());
    }

    public ResponseEntity<CarDto> getCarById(Long id) {
        return carRepository.findById(id)
                .map(car -> ResponseEntity.ok(CarMapper.toDto(car)))
                .orElseGet(() -> ResponseEntity.notFound().build());
    }

    public ResponseEntity<CarDto> createCar(CarDto carDto) {
        if (carDto.getClientId() == null || !clientRepository.existsById(carDto.getClientId())) {
            return ResponseEntity.badRequest().body(null);
        }

        Car car = CarMapper.toEntity(carDto);
        clientRepository.findById(carDto.getClientId()).ifPresent(car::setClient);
        Car savedCar = carRepository.save(car);
        return ResponseEntity.ok(CarMapper.toDto(savedCar));
    }

    public ResponseEntity<CarDto> updateCar(Long id, CarDto carDto) {
        Optional<Car> carOptional = carRepository.findById(id);
        if (carOptional.isPresent()) {
            Car carToUpdate = carOptional.get();

            if (carDto.getVin() != null) {
                carToUpdate.setVin(carDto.getVin());
            }
            if (carDto.getMarkModel() != null) {
                carToUpdate.setMarkModel(carDto.getMarkModel());
            }
            if (carDto.getYearManufacture() != 0) {
                carToUpdate.setYearManufacture(carDto.getYearManufacture());
            }
            if (carDto.getPlateNum() != null) {
                carToUpdate.setPlateNum(carDto.getPlateNum());
            }
            if (carDto.getLastServiceDate() != null) {
                carToUpdate.setLastServiceDate(carDto.getLastServiceDate());
            }

            if (carDto.getClientId() != null && (carToUpdate.getClient() == null || !carToUpdate.getClient().getId().equals(carDto.getClientId()))) {
                Optional<Client> clientOptional = clientRepository.findById(carDto.getClientId());
                if (clientOptional.isEmpty()) {
                    return ResponseEntity.badRequest().build();
                }
                carToUpdate.setClient(clientOptional.get());
            }

            Car updatedCar = carRepository.save(carToUpdate);
            return ResponseEntity.ok(CarMapper.toDto(updatedCar));
        } else {
            return ResponseEntity.notFound().build();
        }
    }


    public ResponseEntity<Void> deleteCar(Long id) {
        if (carRepository.existsById(id)) {
            carRepository.deleteById(id);
            return ResponseEntity.noContent().build();
        } else {
            return ResponseEntity.notFound().build();
        }
    }

    public List<CarDto> searchCars(Long carId, String vin, String markModel, String plateNum, Integer yearManufacture, String lastServiceDate, Long clientId) {
        List<Car> cars;
        if (carId != null) {
            Optional<Car> carOpt = carRepository.findById(carId);
            if (carOpt.isPresent()) {
                cars = List.of(carOpt.get());
            } else {
                cars = Collections.emptyList();
            }
        }else if (vin != null) {
            cars = carRepository.findByVinContainingIgnoreCase(vin);
        } else if (markModel != null) {
            cars = carRepository.findByMarkModelContainingIgnoreCase(markModel);
        } else if (plateNum != null) {
            cars = carRepository.findByPlateNumContainingIgnoreCase(plateNum);
        } else if (yearManufacture != null) {
            cars = carRepository.findByYearManufacture(yearManufacture);
        } else if (lastServiceDate != null) {
            cars = carRepository.findByLastServiceDate(LocalDate.parse(lastServiceDate));
        } else if (clientId != null) {
            cars = carRepository.findByClientId(clientId, Sort.by(Sort.Direction.ASC, "id"));
        } else {
            cars = carRepository.findAll();
        }

        return cars.stream()
                .map(CarMapper::toDto)
                .toList();
    }


}

File: frontend/src/app/car/car.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface CarDto {
  id?: number;
  vin?: string;
  markModel?: string;
  yearManufacture?: number;
  plateNum?: string;
  lastServiceDate?: string;
  clientId?: number;
}

@Injectable({
  providedIn: 'root'
})
export class CarService {
  private baseUrl = 'http://localhost:8080/cars';

  constructor(private http: HttpClient) {}

  getAllCars(): Observable<CarDto[]> {
    return this.http.get<CarDto[]>(this.baseUrl);
  }

  getCarById(id: number): Observable<CarDto> {
    return this.http.get<CarDto>(`${this.baseUrl}/${id}`);
  }

  createCar(car: CarDto): Observable<CarDto> {
    return this.http.post<CarDto>(this.baseUrl, car);
  }

  updateCar(id: number, car: CarDto): Observable<CarDto> {
    return this.http.put<CarDto>(`${this.baseUrl}/${id}`, car);
  }

  deleteCar(id: number): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/${id}`);
  }

  searchCars(params: {
    carId?: number;
    vin?: string;
    markModel?: string;
    yearManufacture?: number;
    plateNum?: string;
    clientId?: number;
    lastServiceDate?: string;
  }): Observable<CarDto[]> {
    const queryParams = new URLSearchParams();
    if (params.carId) queryParams.append('carId', params.carId.toString());
    if (params.vin) queryParams.append('vin', params.vin);
    if (params.markModel) queryParams.append('markModel', params.markModel);
    if (params.yearManufacture) queryParams.append('yearManufacture', params.yearManufacture.toString());
    if (params.plateNum) queryParams.append('plateNum', params.plateNum);
    if (params.clientId) queryParams.append('clientId', params.clientId.toString());
    if (params.lastServiceDate) queryParams.append('lastServiceDate', params.lastServiceDate);

    const queryString = queryParams.toString();
    const url = queryString ? `${this.baseUrl}/search?${queryString}` : this.baseUrl;

    return this.http.get<CarDto[]>(url);
  }
}
